### ISSUE TITLE
Clarification of CRAM mate record resolution

### ISSUE THREAD

**Issue by cmnbroad** [samtools/hts-specs#458]
@jkbonfield I'm trying to understand how to encode (to CRAM) and then subsequently recover the mate graph for a given set SAM records (abbreviated here - I can supply a test file if necessary):

HK35MCCXX160204:8:2209:4005:44116       99      20      7000    0       151M    =       7173    281...
HK35MCCXX160204:8:2209:4005:44116       2179    20      7172    0       44M107S =       7000    -281...
HK35MCCXX160204:8:2209:4005:44116       147     20      7173    0       108M43S =       7000    -281...

You can see from both the bam flags and the mate alignment start that the first record is mated to the third one, with the middle one being supplementary. I would expect that a CRAM encoding of the first one would (or at least could) use CRAM 0x4 flag (mate downstream), with NF=2. 

When I use samtools to write these to CRAM, the first one is encoded with the CRAM 0x4 flag and NF=1 (incorrect in my mind), but when reconverting the CRAM back to SAM/BAM, samtools correctly recovers the pairing and restores these back to their original order and relationship. htsjdk restores them as well, but only due to what I consider to be a bug, with this outcome as an accidental side effect. So, am I misinterpreting how NF is interpreted, or is NF=1 correct in this case ?

**Comment by jkbonfield** [comment]
If you just have the 1st and 3rd of these and replace that supplementary with another template name, then we get this (edited from `cram_dump` output)

```
Rec 0:
BF = 99
CF = 5 (Not detached, and mate is downstream)
RN = HK35MCCXX160204:8:2209:4005:44116
NF = 1+1 = 2

Rec 1: irrelevant as different template

Rec 2:
BF = 147
CF = 1
RN = HK35MCCXX160204:8:2209:4005:44116
```

Basically it's detected the primary pair can be reconstructed so it's linked record 0 to record 2

If we use the middle read as a supplementary being in this same template then it turns into this:

```
Rec 0:
BF = 99
CF = 5 (Not detached, and mate is downstream)
RN = HK35MCCXX160204:8:2209:4005:44116
NF = 1+1 = 2

Rec 1:
BF = 2179
CF = 3 (Detached)
RN = HK35MCCXX160204:8:2209:4005:44116 (ret 0, out_sz 33)
MF = 0
NS = 0
NP = 7000
TS = -281

Rec 2:
BF = 147
CF = 1
RN = HK35MCCXX160204:8:2209:4005:44116
```

You'll see records 0 and 2 haven't changed which is correct given they are the primary reads and they are still pointing towards one another.  The additional supplementary read doesn't fit well into this, so it has the data stored verbatim.

**Comment by jkbonfield** [comment]
Maybe this is just confusion over counting - from 0 or from 1.  The pseudocode isn't complete here as it's such a nightmarish thing to describe algorithmically.   We just say "Resolve ..." with `this` and `this + next_frag` records.

I think it's basically the number of additional records to skip, hence it's a fence-post type problem of counting panels vs posts.  Both C and Java agree on this, although my code would simply have been copied from the Java implementation as this pairing stuff was done before I got involved.

**Comment by jkbonfield** [comment]
I believe this was addressed in #462.
